lib/src/ssh_signing.rs

```diff
@@ -0,0 +1,305 @@
+    let mut file = File::open(path).unwrap();
+    file.read_to_end(&mut buf).unwrap();
+
+    assert_eq!("some-key-data", String::from_utf8(buf).unwrap());
+}
+
+#[test]
+fn test_ssh_key_to_file_conversion_existing_file() {
+    use std::io::Write;
+
+    let mut file = tempfile::Builder::new()
+        .prefix("jj-signing-key-")
+        .tempfile()
+        .map_err(SshError::Io)
+        .unwrap();
+
+    file.write_all(b"some-data").map_err(SshError::Io).unwrap();
+    file.flush().map_err(SshError::Io).unwrap();
+
+    let file_path = file.into_temp_path();
+
+    let (path, file) = ensure_key_as_file(file_path.to_str().unwrap()).unwrap();
+
+    assert_eq!(file_path.to_str().unwrap(), path.to_str().unwrap());
+    assert!(file.is_none());
+}
+
+impl SshBackend {
+    pub fn new(program: OsString, allowed_signers: Option<OsString>) -> Self {
+        Self {
+            program,
+            allowed_signers,
+        }
+    }
+
+    pub fn from_config(config: &config::Config) -> Self {
+        Self::new(
+            config
+                .get_string("signing.backends.ssh.program")
+                .unwrap_or_else(|_| "ssh-keygen".into())
+                .into(),
+            config
+                .get_string("signing.backends.ssh.allowed-signers")
+                .map_or_else(|_| None, |value| Some(value.into())),
+        )
+    }
+
+    fn create_command(&self) -> Command {
+        let mut command = Command::new(&self.program);
+
+        command
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped());
+
+        command
+    }
+
+    fn get_allowed_signers(&self) -> Result<OsString, SshError> {
+        if let Some(allowed_signers) = &self.allowed_signers {
+            Ok(allowed_signers.into())
+        } else {
+            Err(SshError::MissingAllowedSigners {})
+        }
+    }
+
+    fn find_principal(
+        &self,
+        signature_file_path: &tempfile::TempPath,
+    ) -> Result<Option<String>, SshError> {
+        let allowed_signers = self.get_allowed_signers()?;
+
+        let mut command = self.create_command();
+
+        command
+            .arg("-Y")
+            .arg("find-principals")
+            .arg("-f")
+            .arg(allowed_signers.as_os_str())
+            .arg("-s")
+            .arg(signature_file_path.as_os_str());
+
+        let process = command.spawn()?;
+        let output = process.wait_with_output()?;
+
+        let principal = parse_utf8_string(output.stdout)?
+            .split('\n')
+            .next()
+            .unwrap()
+            .trim()
+            .to_string();
+
+        if principal.is_empty() {
+            return Ok(None);
+        }
+        Ok(Some(principal))
+    }
+}
+
+impl SigningBackend for SshBackend {
+    fn name(&self) -> &str {
+        "ssh"
+    }
+
+    fn can_read(&self, signature: &[u8]) -> bool {
+        signature.starts_with(b"-----BEGIN SSH SIGNATURE-----") && self.allowed_signers.is_some()
```

Comment:
Similar to what Yuja said in other comment, the `&& self.allowed_signers.is_some()` part should not be there - `can_read` should be static (it has &self only because SshBackend has to be a trait object).

In the verify method you could short-circuit and return `Verification::Unknown` if there are no allowed signers instead - this feels just more consistent at least

Reply:
Ah good catch this was actually a hack I had added while I was using this branch locally to disable showing verifications in the log. I was using this before I had implemented the `show-signatures` config option.

I have removed this now.